RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,566 @@
+
+Template state would look like:
+
+``` ruby
+class Template
+  attr_reader :name            # The name of the template
+  attr_reader :path            # On disk path to the template files
+  attr_reader :state           # [TemplateState] The current state of the template
+end
+
+class TemplateState
+  attr_reader :template        # [Template] The template this is the state for
+  attr_reader :settings        # A ruby hash of the settings
+  attr_reader :files_to_render # An array of strings. Relative path of files to render
+end
+```
+
+And a lifecycle hook file may look like:
+
+``` ruby
+#
+# after_project_settings.rb
+#
+def after_project_settings(template_state)
+  if template_state.template.name == 'foo'
+    template_state.settings['bar'] = 'baz'
+  else
+    template_state.settings['bar'] = 'Nope!'
+  end
+
+  template_state
+end
+```
+
+### File Contents
+
+#### `<name>/template.json`
+
+| JSON Element | Description |
+| ------------ | ----------- |
+| name (required) | A short descriptive name of the template |
+| description | A long (even multiline) description of the template. Default is the name of the template |
+| type | What type of item this is a template for. Currently only `module` is supported, but in time `control_repo` and others may be added. Default is `module` |
+| default_settings | A YAML hash of the default settings for the template. Default is empty hash |
+| always_apply | Always apply this template. If not specified in the template list, it will be applied last. Default is `false` |
+| tags | An array of string tags for this template. Always includes the directory name of the module. Default is an empty array |
+
+Example for the PDK Ignore template.
+
+This template manages the `.pdkignore` file.  It subscribes to the `use_litmus` setting which will then add `/inventory.yml` to its paths
+
+``` json
+{
+  "name": "PDK Ignore",
+  "description": "The PDK Ignore template manages the .pdkignore file for Puppet Modules",
+  "type": "module",
+  "tags": [
+    "ignore",
+    "default",
+    "pdkignore",
+  ],
+  "always_apply": true,
+  "setting_subscriptions": [
+    "use_litmus"
+  ],
+  "default_settings": {
+    "paths": [
+      "/appveyor.yml",
+      "/.fixtures.yml",
+      "/Gemfile",
+      "/.gitattributes",
+      "/.gitignore",
+      "/.gitlab-ci.yml",
+      "/.pdkignore",
+      "/Rakefile",
+      "/rakelib/",
+      "/.rspec",
+      "/.rubocop.yml",
+      "/.travis.yml",
+      "/.yardopts",
+      "/spec/",
+      "/.vscode/"
+    ]
+  }
+}
+```
+
+``` json
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://puppet.com/schema/does_not_exist.json",
+  "type": "object",
+  "title": "The PDK Ignore Template Schema",
+  "properties": {
+    "use_litmus": {
+      "$id": "#/properties/use_litmus",
+      "title": "Shared setting to assert that Litmus is being used",
+      "description": "Shared setting to assert that Litmus is being used",
+      "type": "boolean"
+    },
```

Comment:
Does this mean I could theoretically override the `use_litmus` setting for this template via `.sync.yml`? Is that desired if so?